### ISSUE TITLE
Leaks memory?

### ISSUE THREAD

**Issue by trizen** [gray/set-product-xs#1]
Hi,

I was looking for XS Perl modules that compute the Cartesian product and I discovered this one. It is indeed very fast, but I think it leaks memory (please correct me if I'm wrong).

Here is real-code example, which ends up using 7GB+ of memory:
```perl
use 5.010;
use strict;
use warnings;

use ntheory qw(:all);
use Set::Product::XS qw(product);

sub l {
    my ($n) = @_;

    # Power of two (n > 4)
    if (not($n & ($n - 1)) and $n > 4) {
        return (($n >> 1) + 1);
    }

    # Prime power or twice a prime power
    if (is_prime_power($n) or (valuation($n, 2) == 1 and is_prime_power($n >> 1))) {
        return 1;
    }

    my %table;
    foreach my $f (factor_exp($n)) {
        my $pp = $f->[0]**$f->[1];

        if ($pp == 2) {
            push(@{$table{$pp}}, [1, $pp]);
        }
        elsif ($pp == 4) {
            push(@{$table{$pp}}, [1, $pp], [3, $pp]);
        }
        elsif ($pp % 2 == 0) {    # 2^k, where k >= 3
            push(@{$table{$pp}},
                [$pp / 2 - 1, $pp], [$pp - 1, $pp],
                [$pp / 2 + 1, $pp], [$pp + 1, $pp]);
        }
        else {                    # odd prime power
            push(@{$table{$pp}}, [1, $pp], [$pp - 1, $pp]);
        }
    }

    my $solution = 1;

    # Generate the solutions and pick the largest one bellow n-1
    product {
        my $x = chinese(@_);
        if ($x > $solution and $x < $n - 1) {
            $solution = $x;
        }
    } values %table;

    return $solution;
}

my $sum = 0;
foreach my $d (3 .. 2e7) {      # ends up using 7GB+ of memory
    $sum += l($d);
}
say $sum;
```

The [original version](https://github.com/trizen/project-euler/blob/f8c66af75b5af02748dcd72c7971485cf925ff43/451%20Modular%20inverses.pl), which uses [Algorithm::Loops](https://metacpan.org/pod/Algorithm::Loops), requires only ~8MB of RAM.

Can you confirm this issue?

Thanks,
Trizen

**Comment by gray** [comment]
It appears you are correct, there is a leak. This simpler test case demonstrates it better:

    #!/usr/bin/env perl
    use strict;
    use warnings;

    # use Set::Product::PP qw(product);
    use Set::Product::XS qw(product);

    for my $d (3 .. 2e7) {
        product {  } [ 1 .. 10 ];
    }

**Comment by gray** [comment]
Fixed in version 0.04, which is now on CPAN.

**Comment by trizen** [comment]
Awesome! Thank you very much.
